dags/inference/configs/maxtext_inference_microbenchmark_gce_config.py

```diff
@@ -0,0 +1,157 @@
+# Copyright 2024 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Utilities to construct configs for maxtext inference microbenchmarks DAG."""
+
+import datetime
+import json
+from typing import Dict
+from xlml.apis import gcp_config, metric_config, task, test_config
+from dags import test_owner
+from dags.multipod.configs import common
+from dags.vm_resource import TpuVersion, Project, RuntimeVersion
+
+PROJECT_NAME = Project.CLOUD_ML_AUTO_SOLUTIONS.value
+RUNTIME_IMAGE = RuntimeVersion.TPU_UBUNTU2204_BASE.value
+GCS_SUBFOLDER_PREFIX = test_owner.Team.INFERENCE.value
+
+
+def get_maxtext_inference_microbenchmark_nightly_config(
+    tpu_version: TpuVersion,
+    tpu_cores: int,
+    tpu_zone: str,
+    time_out_in_min: int,
+    test_name: str,
+    test_mode: common.SetupMode,
+    project_name: str = PROJECT_NAME,
+    runtime_version: str = RUNTIME_IMAGE,
+    network: str = "default",
+    subnetwork: str = "default",
+    is_tpu_reserved: bool = True,
+    num_slices: int = 1,
+    model_configs: Dict = {},
+) -> task.TpuQueuedResourceTask:
+  job_gcp_config = gcp_config.GCPConfig(
+      project_name=project_name,
+      zone=tpu_zone,
+      dataset_name=metric_config.DatasetOption.BENCHMARK_DATASET,
+  )
+
+  set_up_cmds = (
+      "pip install --upgrade pip",
+      # Download maxtext
+      "git clone -b mor--kv-cache-layout-reformat-output https://github.com/google/maxtext.git",
+      # Create a python virtual environment
+      "sudo apt-get -y update",
+      "sudo apt-get -y install python3.10-venv",
+      "sudo apt-get -y install jq",
+      "python -m venv .env",
+      "source .env/bin/activate",
+      # Setup MaxText
+      f"cd maxtext && bash setup.sh MODE={test_mode.value} && cd ..",
+      "pip install torch --index-url https://download.pytorch.org/whl/cpu",
+  )
+
+  additional_metadata_dict = {
```

Comment:
so all these is already logged, see this comment: https://github.com/google/maxtext/pull/662/files#r1610455454

Instead of log it manually, pass in to maxtext, and log there manually again, which is very error prune, and we will miss when there is new flag added/removed, we can utilize the existing automatically logged file on gcs, or local, just copy those value into dimension_json.

Reply:
Each config has over 200 keys; do we want to log every one of the keys and its associated values to BQ? Should be pretty easy to do
```
import yaml
import json

with open('file.yaml') as f:
    print(json.dumps(yaml.load(f)))
```
But one issue is that there are some entries that are multi-level. i.e.
```
logical_axis_rules: !!python/tuple
- !!python/tuple
  - activation_batch
  - !!python/tuple
    - data
    - fsdp
    - fsdp_transpose
- !!python/tuple
  - activation_heads
  - !!python/tuple
    - tensor
    - sequence
```
XLML needs the `metrics` and `dimensions` dict to be single-level. Not sure the easiest way to compress it to a single-level. One solution is that we could ignore keys like this.

OR, we don't record all of these, and only the most important ones like
```
    dimensions_json = {
      "base_output_directory": config.base_output_directory,
      "model_name": config.model_name,
      "tokenizer": config.tokenizer,
      "weight_dtype": config.weight_dtype,
      "inference_microbenchmark_prefill_lengths": config.inference_microbenchmark_prefill_lengths,
      "inference_microbenchmark_stages": config.inference_microbenchmark_stages,
      "inference_microbenchmark_loop_iters": config.inference_microbenchmark_loop_iters,
      "max_prefill_predict_length": config.max_prefill_predict_length,
      "max_target_length": config.max_target_length,
      "per_device_batch_size": config.per_device_batch_size,
      "ici_fsdp_parallelism": config.ici_fsdp_parallelism,
      "ici_autoregressive_parallelism": config.ici_autoregressive_parallelism,
      "ici_tensor_parallelism": config.ici_tensor_parallelism,
      "enable_profiler": config.enable_profiler,
      "scan_layers": config.scan_layers,
      "quantization": config.quantization,
      "quantize_kvcache": config.quantize_kvcache,
      "attention": config.attention,
      "key_value_axis_order_product_id_list": config.key_value_axis_order_product_id_list,
      "prefill_key_axis_order_list": config.prefill_key_axis_order_list,
      "prefill_value_axis_order_list": config.prefill_value_axis_order_list,
      "ar_key_axis_order_list": config.ar_key_axis_order_list,
      "ar_value_axis_order_list": config.ar_value_axis_order_list,
    }
    dimensions_json = {
      **dimensions_json,
      **inference_metadata,
    }
```
WDYT?

Reply:
We should record the most important ones as separate keys. Can we dump the entire serialized-config-json into a separate entry for debugging? I know its not queryable in BigQuery, but its available for anyone to analysis/debug.

Reply:
Done!